src/components/DnsSettings.tsx

```diff
@@ -1,313 +1,239 @@
+    recordType: settings.dns?.recordType ?? DnsRecordType.A,
+    server: settings!.dns?.server ?? '8.8.8.8',
+    ipVersion: settings!.dns?.ipVersion ?? IpVersion.V4,
+    protocol: settings!.dns?.protocol ?? DnsProtocol.UDP,
+    port: settings!.dns?.port ?? 53,
 
-  onValidateAdditionalChange = (validations: DNSRRValidator | undefined) => {
-    if (!validations) {
-      this.setState({ validateAdditionalRRS: undefined }, this.onUpdate);
-      return;
-    }
-    const validateAdditionalRRS = {
-      failIfMatchesRegexp: validations.failIfMatchesRegexp,
-      failIfNotMatchesRegexp: validations.failIfNotMatchesRegexp,
-    };
-    this.setState({ validateAdditionalRRS }, this.onUpdate);
+    // validation
+    validRCodes: settings!.dns?.validRCodes ?? [DnsResponseCodes.NOERROR],
+    validateAnswerRRS: settings!.dns?.validateAnswerRRS ?? { failIfMatchesRegexp: [], failIfNotMatchesRegexp: [] },
+    validateAuthorityRRS: settings!.dns?.validateAuthorityRRS ?? {
+      failIfMatchesRegexp: [],
+      failIfNotMatchesRegexp: [],
+    },
+    validateAdditionalRRS: settings!.dns?.validateAdditionalRRS ?? {
+      failIfMatchesRegexp: [],
+      failIfNotMatchesRegexp: [],
+    },
   };
-
-  render() {
-    const {
-      ipVersion,
-      recordType,
-      server,
-      port,
-      protocol,
-      showAdvanced,
-      showValidation,
-      validRCodes,
-      validateAnswerRRS,
-      validateAuthorityRRS,
-      validateAdditionalRRS,
-    } = this.state;
-    const { isEditor } = this.props;
-
-    const recordTypes = [
-      {
-        label: DnsRecordType.A,
-        value: DnsRecordType.A,
-      },
-      {
-        label: DnsRecordType.AAAA,
-        value: DnsRecordType.AAAA,
-      },
-      {
-        label: DnsRecordType.CNAME,
-        value: DnsRecordType.CNAME,
-      },
-      {
-        label: DnsRecordType.MX,
-        value: DnsRecordType.MX,
-      },
-      {
-        label: DnsRecordType.NS,
-        value: DnsRecordType.NS,
-      },
-      {
-        label: DnsRecordType.SOA,
-        value: DnsRecordType.SOA,
-      },
-      {
-        label: DnsRecordType.TXT,
-        value: DnsRecordType.TXT,
-      },
-      {
-        label: DnsRecordType.PTR,
-        value: DnsRecordType.PTR,
-      },
-      {
-        label: DnsRecordType.SRV,
-        value: DnsRecordType.SRV,
-      },
-    ];
-    const protocols = [
-      {
-        label: DnsProtocol.UDP,
-        value: DnsProtocol.UDP,
-      },
-      {
-        label: DnsProtocol.TCP,
-        value: DnsProtocol.TCP,
-      },
-    ];
-    const rCodes = [
-      {
-        label: 'NOERROR',
-        value: 'NOERROR',
-      },
-    ];
-    return (
-      <Container>
+  const [dnsSettings, dispatch] = useReducer(dnsSettingsReducer, initialValues);
+  const [showValidation, setShowValidation] = useState(false);
+  const [showAdvanced, setShowAdvanced] = useState(false);
+
+  useEffect(() => {
+    onUpdate({ dns: dnsSettings });
+  }, [dnsSettings]);
```

Comment:
Will this work for the deeply nested values? Afaik shallow comparison is used for hooks dependencies.

Reply:
Good question! I was banking on referential equality working for me here since the reducer always returns a new copy, but that might not be the best idea

Reply:
There are some utils for deep dependency comparison but I'm not sure about performance implications of those. 

I might be missing some context but I'm not sure why do we need to store local state at all and not read it straight from the settings and modify them via `onUpdate`?

Reply:
That's a good question (also don't have a ton of context yet). I think that would work, I'll tinker and see what happens. I think migrating to using Form state would make this all go away anyways

Reply:
Good suggestion! Much cleaner this way